pkg/api/resource/typename.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// Typename is a generic representation for all compute resource typenames.
+// It helps enforce a specific format for naming compute resource types by providing
+// convinient input parsing methods.
+//
+// The naming format for a compute resource type is described below:
+//
+// A fully-qualified compute resource typename is constructed from a
+// DNS-style subdomain with at least one dot, a slash /, and
+// a path comprised of one or more segments separated by slashes.
+// Specifically,
+// 1. The subdomain must conform to RFC 1035 section 2.3.1 'subdomain' syntax
```

Comment:
Yeah, we should just reduce this all to the format of "qualified name" - "kubernetes.io/memory".  I can pull together a cleanup PR or you can do it - your choice.